core/Updates/5.0.0-b1.php

```diff
@@ -110,4 +117,39 @@ private function requiresUpdatedLogVisitTableIndex()
 
         return true;
     }
+
+    /**
+     * Add new 'pages before column' to log conversions
+     *
+     * @return DbAlias\AddColumns
+     */
+    private function getPagesBeforeAddColumn()
+    {
+        return $this->migration->db->addColumns('log_conversion', ['pageviews_before' => "SMALLINT UNSIGNED DEFAULT NULL"]);
+    }
+
+    private function populatePagesBefore(): void
+    {
+        // Abort if host is *.matomo.cloud or *.innocraft.cloud
+        $instanceId = SettingsPiwik::getPiwikInstanceId();
+        if (strpos($instanceId, '.matomo.cloud') !== false || strpos($instanceId, '.innocraft.cloud') !== false) {
+            return;
+        }
+
+        // Abort if there are more than 10,000 conversions across all sites and goals in the last 48hrs
+        $startDate = Date::factory('yesterday');
+        $endDate = Date::factory('today')->getEndOfDay();
+
+        $sql = 'SELECT COUNT(*) FROM ' . Common::prefixTable('log_conversion') . ' WHERE server_time <= ? AND server_time >= ?';
+
+        $result = Db::fetchOne($sql, [$startDate, $endDate]);
+        if ($result > 10000) {
```

Comment:
Since all the new values are calculated for each goal using a single query perhaps the query performance should be tested on a large dataset to get some idea of performance? If it's able to do ~1,000,000 conversions in a reasonable time then perhaps there is no need for a limit or the limit could be increased?

```SQL
UPDATE log_conversion c
LEFT JOIN (
     SELECT COUNT(va.idvisit) AS pages_before, va.idvisit, va.server_time
     FROM log_link_visit_action va
     LEFT JOIN log_action a ON a.idaction = va.idaction_url AND a.type = 1
     GROUP BY va.idvisit
) AS a ON a.idvisit = c.idvisit AND a.server_time <= c.server_time
SET c.pageviews_before = a.pages_before
WHERE c.idsite = ? AND c.idgoal = ? AND c.pageviews_before IS NULL;
```

Reply:
@tsteur To test this I generated a test dataset locally with 3,000,000 conversion records, each with 20 linked visit actions before conversion. Then I ran the update query progressively decreasing the dataset size. 

| Conversions | Query time |
| ---: | ---:|
|3,000,000 |181.2s |
|1,000,000 | 58.9s |
|300,000  |24.1s |
|100,000 | 9.6s. |
|10,000 |1.6s |

That seems fairly quick. 

How long do we consider a reasonable time to be during an update? 
We could also a separate limit lower for cloud and a higher limit (or perhaps none?) for on- premise?

Reply:
@bx80 On Cloud we would execute this update query after the deploy of Matomo 5 for it to process everything else quickly. Meaning we would ideally have a custom command that we could execute that only executes say the most recent X conversions (say the most recent 50K conversions).

Reply:
@tsteur The `core:calculate-conversion-pages` command included in this PR has a date range option, but I can add another option to process just the last X conversions instead. :+1: 

So to summarize, the migration script will not attempt to update any existing conversions for cloud domains, and the new command to update the last X conversions can be run separately after deployment by the cloud team.

Reply:
Awesome, that sounds great @bx80